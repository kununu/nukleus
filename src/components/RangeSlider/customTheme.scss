@import '~@kununu/kununu-theme-v2/scss/base/variables';

$height: 30px;
$thumb-height: 25px;
$track-height: 4px;

$upper-color: lighten($red, 40);
$lower-color: $blue;
$thumb-color: $red;
$thumb-hover-color: $blue;

$upper-background: linear-gradient(to top, $upper-color, $upper-color) 100% 50% / 100% $track-height no-repeat transparent;
$lower-background: linear-gradient(to top, $lower-color, $lower-color) 100% 50% / 100% $track-height no-repeat transparent;

// Webkit cannot style progress so we fake it with a long shadow on the thumb element
@function webkit-slider-thumb-shadow($i: 1) {
  $val: #{$i}px 0 0 -#{($thumb-height - $track-height) / 2} #{$upper-color};
  @if $i == 1 {
    @for $k from 2 through 2000 {
      $val: #{$val}, webkit-slider-thumb-shadow($k);
    }
  }
  @return $val;
}

@mixin thumb {
  height: $thumb-height;
  width: $thumb-height;
  background: #fff;
  border-radius: 100%;
  border: 7px solid $thumb-color;
}

@mixin hover-effect {
  transition: border .2s ease-in-out;
  border-color: darken($thumb-color, 10);
}

.slider {
  appearance: none; // NOSONAR
  -webkit-appearance: none; // NOSONAR
  overflow: hidden;
  cursor: pointer;
  height: 26px;
  margin: 7px 0 10px 0;

  // fix slider height on IE11
  @media all and (-ms-high-contrast: none), (-ms-high-contrast: active) {
    height: 74px;
    margin-top: -8px;
  }

  &:focus {
    position: relative;
    outline: none;

    &::-webkit-slider-thumb {
      @include hover-effect;
      background-color: $upper-color;
    }

    &::-moz-range-thumb {
      @include hover-effect;
      background-color: $upper-color;
    }

    &::-ms-thumb {
      @include hover-effect;
      background-color: $upper-color;
    }
  }

  &::-moz-focus-inner {
    outline: 0 none;
    border: 0;
  }
}

// Webkit
.slider::-webkit-slider-runnable-track {
  width: 100%;
  height: $height;
  background: $lower-background;
}

.slider::-webkit-slider-thumb {
  @include thumb;
  position: relative;
  appearance: none; // NOSONAR
  -webkit-appearance: none; // NOSONAR

  top: 50%;
  margin-top: (-$thumb-height/2);
  box-shadow: webkit-slider-thumb-shadow();
}

// Firefox
.slider::-moz-range-track,
.slider::-moz-range-progress {
  width: 100%;
  height: $height;
  background: $upper-background;
}

.slider::-moz-range-progress {
  background: $lower-background;
}

.slider::-moz-range-thumb {
  @include thumb;
  appearance: none; // NOSONAR
  margin: 0;
  box-sizing: border-box;
  display: block;
}

// Internet Exploder
.slider::-ms-track {
  width: 100%;
  height: $height;
  border: 0;
  // color needed to hide track marks
  color: transparent;
  background: transparent;
}

.slider::-ms-fill-lower {
  background: $lower-background;
}

.slider::-ms-fill-upper {
  background: $upper-background;
}

.slider::-ms-thumb {
  box-sizing: border-box;
  appearance: none; // NOSONAR
  height: $thumb-height;
  width: $thumb-height;
  border-radius: 100%;
  transition: background-color 150ms;
  // IE Edge thinks it can support -webkit prefixes
  top: 0;
  margin: 0;
  box-shadow: none;

  background: #fff;
  border: 7px solid $thumb-color;
}

// dont show tooltip above the slider handle in IE
.slider::-ms-tooltip {
  display: none;
}

.slider:hover {
  &::-webkit-slider-thumb {
    @include hover-effect;
    position: relative;
  }

  &::-moz-range-thumb {
    @include hover-effect;
  }

  &::-ms-thumb {
    @include hover-effect;
  }
}


.sliderGridContainer {
  width: 100%;
  position: relative;
  display: grid;
  grid-template-columns: 40px auto 40px;
}

.rangeSliderWrapper {
  padding: 0 30px;
  width: 100%;

  @supports (display: grid) {
    padding: 0;
  }
}

.modifyAmountBtn {
  @supports (display: grid) {
    position: relative;
  }

  display: block;
  width: 35px;
  height: 40px;
  margin-top: 2px;
  background: none;
  border: none;
  padding: 0;
  text-align: left;

  // fix for IOS9 and IE11
  position: absolute;
  left: 0;

  &:last-of-type {
    margin-left: auto;
    text-align: right;

    // fix for IOS9 and IE11
    top: 0;
    right: 0;
  }

  svg {
    margin-left: auto;
  }

  &:focus {
    outline: none;

    circle {
      fill: $upper-color;
    }
  }
}
