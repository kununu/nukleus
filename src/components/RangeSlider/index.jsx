import React from 'react';
import PropTypes from 'prop-types';

import styles from './index.scss';

export default class RangeSlider extends React.Component {
  static propTypes = {
    min: PropTypes.number.isRequired,
    max: PropTypes.number.isRequired,
    step: PropTypes.number.isRequired,
    onChange: PropTypes.func,
  };

  static defaultProps = {
    onChange: () => {},
  };

  state = {
    value: 10000,
  }

  handleInput = (e) => {
    // set value on the label above the slider
    const {onChange} = this.props;

    this.setState({value: Number(e.target.value)});
    onChange();
  };

  render () {
    const {step, min, max} = this.props;
    const {value} = this.state;
    const labelValue = `${value.toLocaleString()} â‚¬`;

    return (
      <div className={styles.sliderWrapper}>
        <label
          htmlFor="range-slider"
          className={styles.sliderLabel}
        >
          {labelValue}
        </label>
        <span className={styles.sliderMinLimit}>{min}</span>
        <input
          id="range-slider"
          type="range"
          step={step}
          min={min}
          max={max}
          name="range-slider"
          value={value}
          className={styles.slider}
          onInput={this.handleInput}
          onChange={this.handleInput}
        />
        <span className={styles.sliderMaxLimit}>{max}</span>
      </div>
    );
  }
}
