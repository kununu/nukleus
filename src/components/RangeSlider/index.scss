@import '~@kununu/kununu-theme-v2/scss/base/variables';

$height: 30px;
$thumb-height: 25px;
$track-height: 5px;

$upper-color: lighten($btn-secondary-bg, 40);
$lower-color: $btn-secondary-bg;
$thumb-color: $btn-secondary-bg;
$thumb-hover-color: $btn-secondary-bg;

$upper-background: linear-gradient(to top, $upper-color, $upper-color) 100% 50% / 100% $track-height no-repeat transparent;
$lower-background: linear-gradient(to top, $lower-color, $lower-color) 100% 50% / 100% $track-height no-repeat transparent;

// Webkit cannot style progress so we fake it with a long shadow on the thumb element
@function webkit-slider-thumb-shadow($i: 1) {
  $val: #{$i}px 0 0 -#{($thumb-height - $track-height) / 2} #{$upper-color};
  @if $i == 1 {
    @for $k from 2 through 1000 {
      $val: #{$val}, webkit-slider-thumb-shadow($k);
    }
  }
  @return $val;
}

@mixin thumb() {
  height: $thumb-height;
  width: $thumb-height;
  background: #fff;
  border-radius: 100%;
  border: 7px solid $thumb-color;
} 

@mixin hoverEffect() {
  transition: border .2s ease-in-out;
  border-color: darken($thumb-color, 10);
}

.sliderWrapper {
  max-width: 768px;
  position: relative;
}

.slider {
  appearance: none;
  -webkit-appearance: none;
  width: 100%;
  height: $height;
  overflow: hidden;
  cursor: pointer;
  
  &:focus {
    outline: none;
  }
}

// Webkit
.slider::-webkit-slider-runnable-track {
  width: 100%;
  height: $height;
  background: $lower-background;
}

.slider::-webkit-slider-thumb {
  position: relative;
  appearance: none;
  -webkit-appearance: none;
  top: 54%;
  margin-top: (-$thumb-height/2);
  box-shadow: webkit-slider-thumb-shadow();
  @include thumb();
}

.safari::-webkit-slider-thumb {
  top: 50%;
}

// Firefox
.slider::-moz-range-track,
.slider::-moz-range-progress {
  width: 100%;
  height: $height;
  background: $upper-background;
}

.slider::-moz-range-progress {
  background: $lower-background;
}

.slider::-moz-range-thumb {
  appearance: none;
  margin: 0;
  box-sizing: border-box;
  @include thumb();
}

// Internet Exploder
.slider::-ms-track {
  width: 100%;
  height: $height;
  border: 0;
  // color needed to hide track marks
  color: transparent;
  background: transparent;
}

.slider::-ms-fill-lower {
  background: $lower-background;
}

.slider::-ms-fill-upper {
  background: $upper-background;
}

.slider::-ms-thumb {
  appearance: none;
  height: $thumb-height;
  width: $thumb-height;
  background: $thumb-color;
  border-radius: 100%;
  border: 0;
  transition: background-color 150ms;
  // IE Edge thinks it can support -webkit prefixes
  top: 0;
  margin: 0;
  box-shadow: none;
}

// dont show tooltip above the slider handle in IE
.slider::-ms-tooltip {
  display: none;
}

.slider:hover {
  &::-webkit-slider-thumb {
    position: relative;
    @include hoverEffect;
  }

  &::-moz-range-thumb {
    @include hoverEffect;
  }

  &::-ms-thumb {
    @include hoverEffect;
  }
}

.slider:focus {
  &::-webkit-slider-thumb {
  }
  &::-moz-range-thumb {
  }
  &::-ms-thumb {
  }
}

.sliderLabel {
  width: 100%;
  display: block;
  text-align: center;
  font-size: 34px;
  margin-bottom:25px;
}
