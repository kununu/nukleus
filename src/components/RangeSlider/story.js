import React, {useState} from 'react';
import {storiesOf} from '@storybook/react';
import {defaultOptions} from 'defaultOptions'; // eslint-disable-line import/no-unresolved
import {withOptions} from '@storybook/addon-options';
import {withInfo} from '@storybook/addon-info';
import {withKnobs, number} from '@storybook/addon-knobs/react';

import RangeSlider from './index';

const RangeSliderExample = () => {
  const initialState = {
    sliderValue: 0,
  };
  const [input, handleInput] = useState(initialState);

  return (
    <RangeSlider
      id="salaryCurrencySlider"
      min={number('sliderMinValue', 0)}
      max={number('sliderMaxValue', 1000)}
      step={number('sliderStep', 10)}
      name="amount"
      onChange={e => handleInput({...input, sliderValue: e.target.value})}
      onBlur={() => { }}
      label={`Drag me! ${input.sliderValue || ''}`}
      meta={{
        touched: false,
        error: false,
        value: Number(input.sliderValue),
      }}
    />
  );
};

storiesOf('RangeSlider', module)
  .addDecorator(withOptions({...defaultOptions}))
  .addDecorator(withKnobs)
  .addDecorator((story, context) => withInfo('Rangeslider component!')(story)(context))
  .add('basics', () => (
    <RangeSliderExample />
  ));
