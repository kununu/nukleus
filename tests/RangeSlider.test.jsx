import React from 'react';
import renderer from 'react-test-renderer';
import {mount} from 'enzyme';
import toJson from 'enzyme-to-json';
import RangeSlider from 'RangeSlider'; // eslint-disable-line import/no-unresolved

test('Renders RangeSlider', () => {
  const component = renderer.create((
    <RangeSlider
      min={10000}
      max={150000}
      step={1000}
      onChange={() => {}}
    />
  ));

  const tree = component.toJSON();

  expect(tree).toMatchSnapshot();
});

test('RangeSlider can handle input', () => {
  const component = mount((
    <RangeSlider
      min={10000}
      max={150000}
      step={1000}
      onChange={() => {}}
    />
  ));

  component.find('#range-slider').simulate('change');

  expect(toJson(component)).toMatchSnapshot();
});
