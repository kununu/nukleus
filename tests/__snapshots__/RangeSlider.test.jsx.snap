// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`RangeSlider can handle input 1`] = `
<RangeSlider
  max={150000}
  min={10000}
  onChange={[Function]}
  step={1000}
>
  <div
    className="sliderWrapper"
  >
    <label
      className="sliderLabel"
      htmlFor="range-slider"
    >
      10,000 €
    </label>
    <span
      className="sliderMinLimit"
    >
      10000
    </span>
    <input
      className="slider"
      id="range-slider"
      max={150000}
      min={10000}
      name="range-slider"
      onChange={[Function]}
      onInput={[Function]}
      step={1000}
      type="range"
      value={10000}
    />
    <span
      className="sliderMaxLimit"
    >
      150000
    </span>
  </div>
</RangeSlider>
`;

exports[`Renders RangeSlider 1`] = `
<div
  className="sliderWrapper"
>
  <label
    className="sliderLabel"
    htmlFor="range-slider"
  >
    10,000 €
  </label>
  <span
    className="sliderMinLimit"
  >
    10000
  </span>
  <input
    className="slider"
    id="range-slider"
    max={150000}
    min={10000}
    name="range-slider"
    onChange={[Function]}
    onInput={[Function]}
    step={1000}
    type="range"
    value={10000}
  />
  <span
    className="sliderMaxLimit"
  >
    150000
  </span>
</div>
`;
